parser/internal/parservX/parser.go

```diff
@@ -818,6 +825,146 @@ func (p *parser) parse() (ast.Expression, error) {
 	return result, nil
 }
 
+func (p *parser) parseFunctionDefinitons() error {
+	functions := make(map[string]*ast.FunctionDefinition)
+
+	for {
+		tok, lit, _ := p.scanIgnoreWhitespace()
+		if tok == ast.Name && lit == "def" {
+			function, err := p.parseFunctionDefinition()
+			if err != nil {
+				return err
+			}
+			functions[function.Name] = function
+		} else {
+			p.unscan()
+			break
+		}
+	}
+
+	p.functions = functions
+	return nil
```

Comment:
Added this in a separate commit, so we can discuss it more easily. 
I removed this check from your suggestion
```
if p.functions == nil {
  return nil
}
```
Why return nil if the functions are nil? 🤔 
In the tests I pass an empty map of functions so it will always be nil in the beginning.